api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirMedicationRequestDaoImpl.java

```diff
@@ -9,14 +9,62 @@
  */
 package org.openmrs.module.fhir2.api.dao.impl;
 
+import ca.uhn.fhir.rest.param.ReferenceAndListParam;
+import ca.uhn.fhir.rest.param.TokenAndListParam;
 import lombok.AccessLevel;
 import lombok.Setter;
+import org.hibernate.Criteria;
 import org.openmrs.DrugOrder;
+import org.openmrs.module.fhir2.FhirConstants;
 import org.openmrs.module.fhir2.api.dao.FhirMedicationRequestDao;
+import org.openmrs.module.fhir2.api.search.param.SearchParameterMap;
 import org.springframework.stereotype.Component;
 
 @Component
 @Setter(AccessLevel.PACKAGE)
 public class FhirMedicationRequestDaoImpl extends BaseFhirDao<DrugOrder> implements FhirMedicationRequestDao {
 	
+	@Override
+	protected void setupSearchParams(Criteria criteria, SearchParameterMap theParams) {
+		theParams.getParameters().forEach(entry -> {
+			switch (entry.getKey()) {
+				case FhirConstants.ENCOUNTER_REFERENCE_SEARCH_HANDLER:
+					entry.getValue().forEach(p -> handleEncounterReference("e", (ReferenceAndListParam) p.getParam())
+					        .ifPresent(c -> criteria.createAlias("encounter", "e").add(c)));
+					break;
+				case FhirConstants.PATIENT_REFERENCE_SEARCH_HANDLER:
+					entry.getValue().forEach(patientReference -> handlePatientReference(criteria,
+					    (ReferenceAndListParam) patientReference.getParam(), "person"));
+					break;
+				case FhirConstants.CODED_SEARCH_HANDLER:
```

Comment:
@ibacher  Need some clarity especially on this case and [this case  ](https://github.com/openmrs/openmrs-module-fhir2/pull/197/files#diff-56bc7e606e3de80b87d4a60451e9134fR46) because I am still struggling to understand the two despite reading up from the [documentation](https://www.hl7.org/fhir/medicationrequest.html#search)

Reply:
@reagan-meant They are intended to cover to slightly different types of cases. So the `CODED_SEARCH_HANDLER` should presumably be used where the `code` search parameter is provided and the `MEDICATION_REFERENCE_SEARCH_HANDLER` where a medication reference is provided. The difference between these is present because [the medication requested can be represented either as a coded concept or as a reference to a Medication resource](https://hl7.org/fhir/medicationrequest-definitions.html#MedicationRequest.medication_x_).

Searches by `code` should look something like this: `http://www.nlm.nih.gov/research/umls/rxnorm|206917` (this is the FHIR representation of the RxNORM code for concept 206917, which is an 800mg oral tablet of ibuprofen). Searches by `medication`, however, will be by a FHIR reference to the medication in question. That means they look like this: `Medication/92035b29-ff19-44f2-a4cb-64beeb36b64d` (this is the FHIR representation for a reference to the Medication—OpenMRS Drug—with a uuid of 92035b29-ff19-44f2-a4cb-64beeb36b64d; this could be the uuid of, say, an 800mg oral tablet of ibuprofen).

The query we want to generate for the former looks something like this:

```sql
select do.*
from drug_order do
    join drug d on do.drug_inventory_id = d.drug_id
    join concept c on d.concept_id = c.concept_id
    join concept_reference_map crm on c.concept_id = crm.concept_id
    join concept_reference_term crt on crm.concept_reference_term_id = crt.concept_reference_term_id
    join concept_source cs on crt.concept_source_id = cs.concept_source_id
    join fhir_concept_source fcs on cs.concept_source_id = fcs.concept_source_id
where fcs.url = 'http://www.nlm.nih.gov/research/umls/rxnorm' and crt.code = '206917'
```

The query we want to generate for the latter looks something like this:

```sql
select do.*
from drug_order do
    join drug d on do.drug_inventory_id = d.drug_id
where d.uuid = '92035b29-ff19-44f2-a4cb-64beeb36b64d'
```

Does that clarify things a bit? Basically, we want to handle the `code` search parameter the way we handle other coded values, but the the `medication` search parameter the way we handle other reference values (like `PARTICIPANT_REFERENCE_SEARCH_HANDLER`)

Reply:
Thanks alot @ibacher , This makes a whole lot of sense now. will have a better PR soon